[Collections] Documentation (#3306)
apple/swift-package-manager: Documentation/PackageCollections.md

```bash
$ swift package-collection add https://www.example.com/packages.json --skip-signature-check
```

For package collections hosted on the web, publishers may ask SwiftPM to [enforce the signature requirement](#protecting-package-collections). If a package collection is
expected to be signed but it isn't, user will see the following error message:

```bash
$ swift package-collection add https://www.example.com/bad-packages.json
The collection is missing required signature, which means it might have been compromised.
```

Users should NOT add the package collection in this case.

##### Trusted root certificates

Since generating a collection signature requires a certificate, part of the signature check involves validating the certificate and its chain and making sure that the root certificate is trusted.

On Apple platforms, all root certificates that come preinstalled with the OS are automatically trusted. Users may include additional certificates to trust by placing 
them in the `~/.swiftpm/config/trust-root-certs` directory. 

> On non-Apple platforms, there are no trusted root certificates by default. Only those found in `~/.swiftpm/config/trust-root-certs` are trusted. This means that 

Reviewer: other than those that are shipped with thew pinning configuration?

#### Requirements on signing certificate

Certificates used for signing package collections must meet the following requirements, which are checked and enforced during signature generation (publishers) and verification (SwiftPM users):
- The timestamp at which signing/verification is done must fall within the signing certificate's validity period.
- The certificate's "Extended Key Usage" extension must include "Code Signing".
- The certificate must use either 256-bit EC (recommended) or 2048-bit RSA key.
- The certificate must not be revoked. The certificate authority must support OCSP, which means the certificate must have the "Certificate Authority Information Access" extension that includes OCSP as a method, specifying the responder's URL.
- The certificate chain is valid and root certificate must be trusted.

Non-expired, non-revoked Apple Distribution certificates from [developer.apple.com](https://developer.apple.com) satisfy all of the criteria above.

##### Trusted root certificates

With the `package-collection-sign` tool, the root certificate provided as input for signing a collection is automatically trusted. When SwiftPM user tries to add the collection, however,

> the root certificate must either be preinstalled with the OS (Apple platforms only) or found in the `~/.swiftpm/config/trust-root-certs` directory, otherwise the 

Reviewer: other than those that are shipped with the pinning configuration?

# Package Collections

Package collections, introduced by [SE-0291](https://github.com/apple/swift-evolution/blob/main/proposals/0291-package-collections.md), are
curated lists of packages and associated metadata that make discovery of existing packages easier. They are authored as static JSON documents 
and can be published to the web or distributed to local file systems. 

## Using package collections

With the `swift package-collection` command-line interface, SwiftPM users can subscribe to package collections. Contents of imported package 

> collections are accessible to any clients of libSwiftPM.

Reviewer 2: (Hrm, not sure about which relative link will work once this is merged.)

> collections are accessible to any clients of libSwiftPM.

Author: Adding a link to the local libSwiftPM docs file

`swift package-collection` has the following subcommands:
- `add`: Add a new collection
- `describe`: Get metadata for a collection or a package included in an imported collection
- `list`: List configured collections
- `refresh`: Refresh configured collections
- `remove`: Remove a configured collection

> - `search`: Search for packages by keywords or module names within imported collections

Author: good idea

### `add` subcommand

This subcommand adds a package collection hosted on the web (HTTPS required):

```bash
$ swift package-collection add https://www.example.com/packages.json
Added "Sample Package Collection" to your package collections.
```

Or found in the local file system:

```bash
$ swift package-collection add file:///absolute/path/to/packages.json
Added "Sample Package Collection" to your package collections.
```

#### Signed package collections

Package collection publishers may sign a collection to protect its contents from being tampered with. If a collection is signed, SwiftPM will check that the 
signature is valid before importing it and return an error if any of these fails:
- The file's contents, signature excluded, must match what was used to generate the signature. In other words, this checks to see if the collection has been altered since it was signed.
- The signing certificate must meet all the [requirements](#requirements-on-signing-certificate).

```bash
$ swift package-collection add https://www.example.com/bad-packages.json
The collection's signature is invalid. If you would still like to add it please rerun 'add' with '--skip-signature-check'.
```

> User may continue adding the collection despite of the error or opt to skip the signature check on a package collection by passing the `--skip-signature-check` flag:

Author: Yep, what you suggested sounds better. Thanks!

On Apple platforms, all root certificates that come preinstalled with the OS are automatically trusted. Users may include additional certificates to trust by placing 
them in the `~/.swiftpm/config/trust-root-certs` directory. 

On non-Apple platforms, there are no trusted root certificates by default. Only those found in `~/.swiftpm/config/trust-root-certs` are trusted. This means that 
the signature check will always fail unless the `trust-root-certs` directory is set up:

```bash
$ swift package-collection add https://www.example.com/packages.json
The collection's signature cannot be verified due to missing configuration.
```

Users should obtain root certificates from collection publishers and save them to `~/.swiftpm/config/trust-root-certs`. The root certificates must be DER-encoded.

#### Unsigned package collections

Users will get an error when trying to add an unsigned package collection:

```bash
$ swift package-collection add https://www.example.com/packages.json
The collection is not signed. If you would still like to add it please rerun 'add' with '--trust-unsigned'.
```

> To continue user must confirm their trust by passing the `--trust-signed` flag:

Reviewer 3: `--trust-unsigned`, right?

Author: Oops thanks. Fixed in https://github.com/apple/swift-package-manager/pull/3306/commits/0a51e58fd95b0718c9bc062b2df7cb3aff03bcfc

```

#### Metadata for a package version

User may view additional metadata for a package version by passing `--version`:

```bash
$ swift package-collection describe --version 4.0.0 https://github.com/jpsim/yams
Package Name: Yams
Version: 4.0.0
Modules: Yams, CYaml
Supported Platforms: iOS, macOS, Linux, tvOS, watchOS
Supported Swift Versions: 5.3, 5.2, 5.1, 5.0
License: MIT
```

### `list` subcommand

This subcommand lists all collections that are configured by the user:

```bash
$ swift package-collection list
Sample Package Collection - https://example.com/packages.json
...
```

### `refresh` subcommand

This subcommand refreshes any cached data manually:

```bash
$ swift package-collection refresh
Refreshed 5 configured package collections.
```

SwiftPM will also automatically refresh data under various conditions, but some queries such as search will rely on locally cached data.

### `remove` subcommand

This subcommand removes as collection from the user's list of configured collections:

```bash
$ swift package-collection remove https://www.example.com/packages.json
Removed "Sample Package Collection" from your package collections.
```

### `search` subcommand

> This subcommand searches for packages by keywords or module names within imported collections.

Reviewer 2: Is there a difference between "configured" and "imported" (or "added", for that matter) collections? Probably not a huge deal, but sticking to a term might improve understanding.

Author: Yes, "configured" means the collection URL is saved the config file; "imported" means the collection is "configured" and its data saved to the database. It's possible to have a collection that is configured but not imported.

Reviewer 2: Ah, great — that makes sense, thanks.

This file is managed through SwiftPM commands and users are not expected to edit it by hand.

---

## Publishing package collections

Package collections can be created and published by anyone. The [swift-package-collection-generator](https://github.com/apple/swift-package-collection-generator) project provides tooling 
intended for package collection publishers:
- [`package-collection-generate`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionGenerator): Generate a package collection given a list of package URLs
- [`package-collection-sign`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionSigner): Sign a package collection
- [`package-collection-validate`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionValidator): Perform basic validations on a package collection
- [`package-collection-diff`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionDiff): Compare two package collections to see if their contents are different 

### Creating package collections

> All package collections must adhere to the [collection data format](../Sources/PackageCollectionsModel/Formats/v1.md) for SwiftPM to be able to consume them. The recommended way
> to create package collections is to use [`package-collection-generate`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionGenerator), 
> otherwise the data models are available through the [`PackageCollectionsModel` module](../Sources/PackageCollectionsModel).

Reviewer 2: Subjective tweak that splits the second sentence:

```suggestion
All package collections must adhere to the [collection data format](../Sources/PackageCollectionsModel/Formats/v1.md) for SwiftPM to be able to consume them. The recommended way
to create package collections is to use [`package-collection-generate`](https://github.com/apple/swift-package-collection-generator/tree/main/Sources/PackageCollectionGenerator). For custom implementations, the data models are available through the [`PackageCollectionsModel` module](../Sources/PackageCollectionsModel).
```

### Protecting package collections

[Signing](#package-collection-signing-optional) can provide some degree of protection on package collections and reduce the risks of their contents being modified by malicious actors, but it doesn't
prevent the following attack vectors:
- **Signature stripping**: This involves attackers removing signature from a signed collection, causing it to be downloaded as an [unsigned collection](#unsigned-package-collections) and bypassing signature check. In this case, publishers should make it known that the collection is signed, and SwiftPM users should abort the `add` operation when the "unsigned" warning appears on a supposedly signed collection.
- **Signature replacement**: Attackers may modify a collection then re-sign it using a different certificate, either pretend to be the same entity or as some other entity, and SwiftPM will accept it as long as the [signature is valid](#signed-package-collections).

To defend against these attacks, SwiftPM has certificate-pinning configuration that allows collection publishers to:

> - Require signature check on their collections - this defends against "signature stripping"
> - Restrict what certificate can be used for signing - this defends against "signature replacement"

Reviewer 2: The most pedantic of nits — use em-dashes instead of hyphens. 😅

```suggestion
- Require signature check on their collections — this defends against "signature stripping".
- Restrict what certificate can be used for signing — this defends against "signature replacement".
```

Author: Heh ok :)

```bash
$ swift package-collection describe --version 4.0.0 https://github.com/jpsim/yams
Package Name: Yams
Version: 4.0.0
Modules: Yams, CYaml
Supported Platforms: iOS, macOS, Linux, tvOS, watchOS
Supported Swift Versions: 5.3, 5.2, 5.1, 5.0
License: MIT
```

### `list` subcommand

This subcommand lists all collections that are configured by the user:

```bash
$ swift package-collection list
Sample Package Collection - https://example.com/packages.json
...
```

### `refresh` subcommand

This subcommand refreshes any cached data manually:

```bash
$ swift package-collection refresh
Refreshed 5 configured package collections.
```

SwiftPM will also automatically refresh data under various conditions, but some queries such as search will rely on locally cached data.

### `remove` subcommand

This subcommand removes as collection from the user's list of configured collections:

```bash
$ swift package-collection remove https://www.example.com/packages.json
Removed "Sample Package Collection" from your package collections.
```

### `search` subcommand

This subcommand searches for packages by keywords or module names within imported collections.

#### String-based search

> The search command does a string-based search when using the `--keyword` option and returns the list of packages that matches the query:

Reviewer 2: Missing the `s` in `--keywords`.

```suggestion
The search command does a string-based search when using the `--keywords` option and returns the list of packages that matches the query:
```

Author: good 👀
